addons/mail/static/src/utils.js

```diff
@@ -56,3 +65,57 @@ export function useAutoScroll(refName) {
         }
     });
 }
+
+export function useHover(refName, callback = () => {}) {
+    const ref = useRef(refName);
+    const state = useState({ isHover: false });
+    function onHover(hovered) {
+        state.isHover = hovered;
+        callback(hovered);
+    }
+    onMounted(() => {
+        ref.el.addEventListener("mouseenter", () => onHover(true), true);
+        ref.el.addEventListener("mouseleave", () => onHover(false), true);
+    });
+    onWillUnmount(() => {
+        ref.el.removeEventListener("mouseenter", () => onHover(true), true);
+        ref.el.removeEventListener("mouseleave", () => onHover(false), true);
+    });
+    return state;
+}
+
+export function useFocus(refName, callback = () => {}) {
+    const ref = useRef(refName);
+    const state = useState({ isFocus: false });
+    function onFocus(focused) {
+        state.isFocus = focused;
+        callback(focused);
+    }
+    onMounted(() => {
+        ref.el.addEventListener("focusin", () => onFocus(true), true);
+        ref.el.addEventListener("focusout", () => onFocus(false), true);
+    });
+    onWillUnmount(() => {
+        ref.el.removeEventListener("focusin", () => onFocus(true), true);
+        ref.el.removeEventListener("focusout", () => onFocus(false), true);
+    });
```

Comment:
I suppose all `onMounted`/`onWillUnmount` can just become `useExternalListener`.
This should lead to less LOCs.

Reply:
The problem with `useExternalListener` is it only tracks on `window` or `document` and not with the `ref`, so the `focus` event will not work and `mouseenter/leave` will be replaced with `mousemove` and I think keeping code like this is better for now

Reply:
`onExternalClick` (in `@mail/new/utils`) is not the same as `useExternalListener` (in `@odoo/owl`).
The former seems to work only on window, while the second is more flexible and you can provide a target.
I suggest using the 2nd one.

Reply:
but that only tracks the `click` event as far as I see?

Reply:
No, it's any event you provide.
Signature is as follow:
```js
useExternalListener(target, eventName, handler, eventParams)
```

So you could have this:
```js
useExternalListener(ref.el, "focusin", () => onFocus(true), { capture: true });
useExternalListener(ref.el, "focusout", () => onFocus(false), { capture: true });
```

Reply:
I suggest defining another `useExternalListener` in `@mail/new/utils`, which is the same as in `@odoo/owl` but the 1st parameter can be function, so that you can use it and reduce amount of LOCs here.